### PR TITLE
feat: initialization and shutdown

### PR DESCRIPTION
**Pull request by toddbaert**
Adds initialization and shutdown discussed in DRAFT client PR, and OFEPs. This is a piece of the (too big) draft PR [here](https://github.com/open-feature/spec/pull/171). The changes are fairly small and non-breaking.

There's a few non-functional changes here (fixes to formatting and minor enhancements to the parser).

Be sure to review `specification.json` for the net changes.

Most of this is implemented in the [web-sdk](https://github.com/open-feature/js-sdk/tree/main/packages/client/src) already. Some conforming (demo) providers can be found in the playground:

- [cloudbees](https://github.com/open-feature/playground/blob/main/packages/web-cloudbees-provider/src/lib/web-cloudbees-provider.ts)
- [flagsmith](https://github.com/open-feature/playground/blob/main/packages/web-flagsmith-provider/src/lib/web-flagsmith-provider.ts)
- [harness](https://github.com/open-feature/playground/blob/main/packages/web-harness-provider/src/lib/web-harness-provider.ts)
- [launchdarkly](https://github.com/open-feature/playground/blob/main/packages/web-launchdarkly-provider/src/lib/web-launchdarkly-provider.ts)
- [split](https://github.com/open-feature/playground/blob/main/packages/web-split-provider/src/lib/web-split-provider.ts)

### CODE REVIEW

File: specification.json
Comment (toddbaert) [inline specification.json:319]: This is not new - git is just seeing it as an add/remove.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [inline specification/sections/01-flag-evaluation.md:149]: I move we try to use Java examples as they are generally well-understood and familiar to most.

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [inline specification/sections/01-flag-evaluation.md:36]: What happens if either `initialize` or `shutdown` fail for some reason?

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: Good question.

I think the provider stays in a bad state (assuming the initialize is necessary for it to function properly) and probably throws during evaluation, and so the SDK will default for all evaluations.

Long term, my thinking is a failed initialize would fire the `ProviderEvents.ERROR` event as well, which would allow application authors to "do something" when this happens by attaching an event handler. I think that even in this case, the provider would still default until some corrective action is taken.

File: specification/sections/01-flag-evaluation.md
Comment (beeme1mr) [inline specification/sections/01-flag-evaluation.md:38]: Do you think we should specify an initialization timeout? I'm sure this could be handled in a provider or userland but it may be worth explicitly defining the maximum amount of time the initialization process will wait.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: I don't think I did a great job explaining my thinking here, so I'll try again :sweat_smile: 

I think that a lot of the things that can "go wrong" during the initialization in the `setProvider` can go wrong outside that context too. Persistent connections can be lost, APIs can become unavailable, etc. Therefor, we already have to handle these situations "out-of-band" of the `setProvider` function anyway, and I think using events is the best approach for this. For this reason, I think it's best to keep things simple here, and basically just log warnings if things go wrong in the `shutdown/initialization` within the `provider mutator`. The provider may attempt to re-establish whatever state is needed to work correctly, or give up if the error is irrecoverable. If it throws, the SDK will default.

That's my line of thinking at the moment.

Comment (weyert) [reply specification/sections/01-flag-evaluation.md:36]: I think the `shutdown` is typically only called when the process exists so I think throwing an event at such a moment would not be really helpful. As don't think you can really act on it at least in Node.js. For example, if you get a SIGKILL signal or something similar.

I assume you are talking when you want to call `shutdown` during the life time of the application. I have to admit I personally haven't found a use case for switching providers in a running application but I think logging a warning works to notify that something went wrong and needs attention.

File: specification/sections/01-flag-evaluation.md
Comment (weyert) [reply specification/sections/01-flag-evaluation.md:38]: Do we want to clarify what the expected order is? Do we want to call `shutdown` on the previous provider before calling `initialise` on the new provider or the other way around?

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: > I think the `shutdown` is typically only called when the process exists so I think throwing an event at such a moment would not be really helpful. As don't think you can really act on it at least in Node.js. For example, if you get a SIGKILL signal or something similar.
> 
> I assume you are talking when you want to call `shutdown` during the life time of the application. I have to admit I personally haven't found a use case for switching providers in a running application but I think logging a warning works to notify that something went wrong and needs attention.

I think we agree. In this model here are 2 places a provider's shutdown is called:

1) when switching away from that provider (a new provider is set)
2) when the API's shutdown is called (this would be used when the application is terminating, as you mention)

Events don't help much in either case for `shutdown`. For errors during/after `initialize`, I think events are important. Events could be important to signal to the `application author` that a connection was lost or some fundamental state of the provider has changed, even long after initialization.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: @weyert I've tried to cover that [here](https://github.com/open-feature/spec/pull/179/files#diff-1443054d965453435ba3da2907dabbc282d2f808c3a64800f10e05fab8351b1aR11-R27).

Basically, this means that the system should never use a provider that hasn't been initialized or shutdown.

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [reply specification/sections/01-flag-evaluation.md:36]: So I think this proposal may be blocked until our events spec is landed.

Reading this: https://github.com/open-feature/spec/pull/171/files#diff-daf30d840f7e7b1752899ac2c4b268af214e6c8e7058549c8d4218abd2e86ba3

It would suggest that event handlers are registered on clients.  This would mean our code would look something like:

```java
OpenFeature api = OpenFeature.getInstance();
Client client = api.getClient('client-name')
client.addHook(ProviderEvent.ERROR, ....)
api.setProvider(new MyProvider())
```

Having to get a client before you've registered a provider seems kinda odd to me?

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [reply specification/sections/01-flag-evaluation.md:38]: What should the behavior be if someone tries to resolve flag values before the provider is in a ready state?

Comment (beeme1mr) [reply specification/sections/01-flag-evaluation.md:38]: I think the default value should be returned with the [error code](https://docs.openfeature.dev/specification/types#error-code) `PROVIDER_NOT_READY`.

Comment (justinabrahms) [inline specification/sections/01-flag-evaluation.md:38]: I tried to implement this in Java. It turns out that it's more complex than you'd expect b/c the Provider doesn't know if it's ready or not, per the spec. Only the global singleton knows.

This means the API will need to manage the lifecycle which is really awkward. I'd like it if we could add something to the spec which says "providers need to be able to tell us what it's current state is on demand via some accessor function".

File: specification.json
Comment (kinyoklion) [inline specification.json:344]: Depending on what we mean by breaking, new methods on an interface may be breaking depending on the language. (For instance the C# SDK uses C# 7.3, which doesn't have default interface implementations, so extending an interface is considered breaking). Though, those languages could likely use two interfaces.

Just highlighting it because it would be easy to accidentally do this in a breaking way.

(Similarly interface extenstion may be seen as a breaking change for testing, because people implement client interfaces and things, and may not have default implementations.)

File: specification/sections/01-flag-evaluation.md
Comment (Kavindu-Dodan) [reply specification/sections/01-flag-evaluation.md:38]: one option is to mandate a `provider state` which is exposed through the accessor. For example, `ready`, `pending` or `error` state, which can be used to react in evaluations or other usages 🤔 

And in case if the provider does not support init and shutdown, the default can be set to `ready`.

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: > I think the default value should be returned with the [error code](https://docs.openfeature.dev/specification/types#error-code) `PROVIDER_NOT_READY`.

This is my feeling as well.

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: > I tried to implement this in Java. It turns out that it's more complex than you'd expect b/c the Provider doesn't know if it's ready or not, per the spec. Only the global singleton knows.
> 
> This means the API will need to manage the lifecycle which is really awkward. I'd like it if we could add something to the spec which says "providers need to be able to tell us what it's current state is on demand via some accessor function".



> one option is to mandate a `provider state` which is exposed through the accessor. For example, `ready`, `pending` or `error` state, which can be used to react in evaluations or other usages thinking
> 
> And in case if the provider does not support init and shutdown, the default can be set to `ready`.

I think I agree. I actually implemented this initially, but then removed it for simplicity... but you're right, it means the API has to track it, which IS awkward. I think it makes sense to add this back to the provider.

File: specification.json
Comment (toddbaert) [reply specification.json:344]: Absolutely. I think SDK implementations will have to be careful in implementing this. As much as possible we'd like to avoid a new major version.

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [reply specification/sections/01-flag-evaluation.md:36]: @toddbaert I was thinking that we'd close out the events spec first, then this one.. not that we'd combine them.

If you expect someone to respond to a provider being in an error state with a log or metric or something, they will need to know about it. What I was attempting to show with the above block of code is that provider registration happens on the API level, but handling the errors that are generated happens at the Client level. This seems incongruous.

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [reply specification/sections/01-flag-evaluation.md:38]: Boolean indicating whether it's ready seems sufficient to implement this.

Comment (Kavindu-Dodan) [reply specification/sections/01-flag-evaluation.md:38]: 👍 for a boolean with a ready state

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: > What I was attempting to show with the above block of code is that provider registration happens on the API level, but handling the errors that are generated happens at the Client level. This seems incongruous.

At first I thought this was an issue as well, but I think that this is a separation we want, TBH. The code setting the provider likely can't really react to such errors, I think it makes sense to have "client type" code react to it... I don't think that (for example) an alternative provider is going to be set because of an error. I think it's more realistic that areas of the app consuming feature flags (using `client` objects) would take some action or revert to some high-level default.

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: @justinabrahms another possibility would be to add global events as well, and perhaps scope them with the same scope identifiers you're proposing in your multi-provider proposal. I'm not convinced this is a better way to go at this point though.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: OK I will add this.

**Review by Kavindu-Dodan — APPROVED** [review]
My only concern is regarding the newly added `initialize` & `shutdown` functions and attempting to adopt SDKs without breaking current implementations. I assume this will be possible for some languages (ex:- Java) but for others (ex:- Go) it can be a breaking change. 

Regardless, approval from my end 

____

Addendum - Investigation for **Go** to have non-breaking changes 

Given that Go does not have default methods, a workaround would be to rely on embedding [1]. Consider the playground example - https://go.dev/play/p/DKv-arO2PeZ

- Introduce another interface - let’s call this **Events**
- Embed  **Events** into current **Provider** contract **FeatureProvider**
- SDK provides a noop implementation for **Events** which can be used by provider implementations

[1] - https://go.dev/doc/effective_go#embedding

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: Added, see [this commit](https://github.com/open-feature/spec/pull/179/commits/164071b5a1ad19553c19156c8a0b68521a9324d9) for the delta. I hope it's clear enough.

@Kavindu-Dodan @justinabrahms

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: Added, see [this commit](https://github.com/open-feature/spec/pull/179/commits/164071b5a1ad19553c19156c8a0b68521a9324d9) for the delta. I hope it's clear enough.

@Kavindu-Dodan @justinabrahms

File: specification.json
Comment (toddbaert) [reply specification.json:344]: @Kavindu-Dodan has some good thoughts about this WRT the Go SDK.

File: specification/sections/01-flag-evaluation.md
Comment (Kavindu-Dodan) [reply specification/sections/01-flag-evaluation.md:38]: Fine from my end :)

File: specification/sections/02-providers.md
Comment (beeme1mr) [inline specification/sections/02-providers.md:212]: Why does this provider implement the `AutoDisposable` method? Is `dispose` not on the `Provider' interface?

Comment (toddbaert) [reply specification/sections/02-providers.md:212]: AutoDisposable is an builtin language interface in Java, like [IDisposable](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=net-7.0) in C#. https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html

We may or may not want to use it, that's up to the SDK implementer, but this is the reason why I don't want to be prescriptive about the naming of this function. I want to give SDK implementors the flexibility to call the method what they want so they can make that decision for their language.

**Review by beeme1mr — APPROVED** [review]
Looks great! It's a small but important feature.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: @Kavindu-Dodan @justinabrahms @beeme1mr

I'm trying to decide exactly what makes more sense here... though it's simple, I'm somewhat unsure.

In the experimental web SDK, the SDK calls the `initialize()` function when the provider is set, and if that succeeds it updates it's own state tracking the provider's readiness (this is the awkwardness @justinabrahms referred to above). I see the point that maintaining this on the provider feels better, but what I'm not sure of is if the SDK should set that value or the provider should. Does it make more sense for the SDK to run the `initialize()` function and then set this field to ready? Or should we leave that field as read-only from outside and rely on the provider author to set it, and specify that the provider should set it during `initialize()`?

Should we specify something like `the intitialize() function will run if the "ready" state is set to false`? I just worry that this will be an easy thing for providers to screw up. I'm wondering if we should actually just maintain the state in the SDK itself after all.

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [reply specification/sections/01-flag-evaluation.md:36]: It seems like we should have some mechanism to detect if there are errors during initialization without requiring a client. But if there is broad disagreement w/ that approach then don't let this be a blocker.

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: @justinabrahms 

> It seems like we should have some mechanism to detect if there are errors during initialization without requiring a client.

I spent some time thinking about this. I think that in most real-world production cases, it's the [`application author`](https://openfeature.dev/specification/glossary#application-author) doing the feature evaluation and dealing with the client abstraction that needs to handle errors (loss of connection etc). The author of that particular part of the application knows how best to handle errors in that area.

That said, there's 2 use-cases I can think of where an [`application integrator`](https://openfeature.dev/specification/glossary#application-integrator) would need similar error signals:

- when initially configuring a provider, it might be tedious to have to get a client to troubleshoot connection issues, authentication issues, etc
- for global monitoring/telemetry purposes (sending a message to sentry or similar, for instance)

So I think I see your point. I have two ideas, which are not mutually exclusive:

1) cause `setProvider` to throw after a (configurable?) timeout, if there's an error in `initialize()`

This would at least help with the first problem. It doesn't really help too much with the second, since setting the provider is a one-time thing, and subsequent errors with the provider aren't captured without attaching a client. I think this would be easy to implement, and it wouldn't violate our "don't throw" rule since that doesn't apply to configuration. Practically, this might constitute a breaking change for JS implementations, since errors here would be associated with asynchronous code, meaning the `setProvider` would need to return a promise.

2) add "global" events. 

These would be similar to the client events, but not scoped to a single provider, wouldn't require the creation of a client. I think we still want most events to be associated with a client (expecially because in a mutli-provider scenario, we need a way to scope events). I think global events could be useful for adding handlers for monitoring the system in it's entirety, and not so much application-level concerns like flag value changes. Put another way, global events would be useful to  [`application integrators`](https://openfeature.dev/specification/glossary#application-integrator) and client events would be useful to [`application authors`]([`application authors`](https://openfeature.dev/specification/glossary#application-author)). We may want to only support error event types in global events.

File: specification/sections/01-flag-evaluation.md
Comment (lukas-reining) [reply specification/sections/01-flag-evaluation.md:38]: Generally it feels better, to have the state of the provider as a property of the provider.
I would totally agree that lowering the coherence of the provider by pulling provider state out of the provider is somewhat awkward.

>  I just worry that this will be an easy thing for providers to screw up. 

This could be mitigated by the SDK implementations. E.g. the default implementation of that method could just be setting the `ready state` to true, so providers would have to overwrite the method and call the super method.
Also It seems to me that specifying that the `ready flag` has to be set to true after running the initialize, should be enough to test providers for their spec compliance.

Comment (lukas-reining) [reply specification/sections/01-flag-evaluation.md:38]: What I was also wondering about is what happens if the client using that provider is doing feature flag evaluation before the `ready flag` is set to true.
I think there is nothing yet defined in the spec right? If thats true I think we should add that to the specification.
I think the spec should mention that in that case a PROVIDER_NOT_READY should be returned.

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: > This could be mitigated by the SDK implementations. E.g. the default implementation of that method could just be setting the `ready flag` to true, so providers would have to overwrite the method and call the super method.

Not all languages have default implementations, but I think there's equivalents and I agree with your point in principle.

> Also It seems to me that specifying that the `ready flag` has to be set to true after running the initialize, should be enough to test providers for their spec compliance.

To be clear: do you feel the SDK should set this, or the provider should? Is this a requirement on the SDK or the provider author?

> I think the spec should mention that in that case a PROVIDER_NOT_READY should be returned.

I think this might be a good idea. Yes. I'll think about that a bit more.

Comment (lukas-reining) [reply specification/sections/01-flag-evaluation.md:38]: > Not all languages have default implementations, but I think there's equivalents and I agree with your point in principle.

Yeah thats true, I just had in mind that in many cases this at least weakens the problem.

> To be clear: do you feel the SDK should set this, or the provider should? Is this a requirement on the SDK or the provider author?

I would say it is a requirement on the provider author, or maybe better the provider implementation.
This also means that if the provider defines that flag, it should always be initially false, which is different from not defining the flag as the spec says that in this case the provider should be considered ready.

> I think this might be a good idea. Yes. I'll think about that a bit more.

Sounds good! 👍

**Review by thomaspoignant — APPROVED** [review]
Super nice addition, this is actually something I already need in the GO SDK.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: > This also means that if the provider defines that flag, it should always be initially false, which is different from not defining the flag as the spec says that in this case the provider should be considered ready.

I think there are providers that are "immediately ready" such as those that pull from environment variables or otherwise get their configuration synchronously from their environment.

Comment (weyert) [reply specification/sections/01-flag-evaluation.md:38]: I think it is reasonable to fallback to the default value of the flag when the provider is not ready yet. And then pass the reason as suggested in the evaluation result.  Then the implementor/consumer of OpenFeature can decide whether the want to postpone execution until the provider is ready or fallback to the default value of the flag?

**Review by weyert — APPROVED** [review]

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:38]: @lukas-reining @weyert thanks!

I've made the changes described above. I've made the language clearer about the fact it's the provider's responsibility to set the ready member, and also added a point about resolution with providers that are not ready. See [delta](https://github.com/open-feature/spec/pull/179/commits/e5ba035ce2f256e8259dd9d9a009123c4d1733c6).

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [reply specification/sections/01-flag-evaluation.md:36]: Global events could be annotated with the provider they reference. This would allow folks to tie into all events at once, which would be useful for the two use-cases mentioned above.

I agree that this is an application integrator concern. I don't have a preference between the two options you have. 2 seems more powerful. 1 seems easier to implement (in java).

File: specification/sections/01-flag-evaluation.md
Comment (justinabrahms) [inline specification/sections/01-flag-evaluation.md:40]:
```suggestion
The `provider's` readiness can state can be determined from it's `ready` member/accessor.
```
Shouldn't be doing field-level accessing in java-land.

File: specification/sections/02-providers.md
Comment (justinabrahms) [inline specification/sections/02-providers.md:193]:
```suggestion
> The `provider` **MAY** define a mutable boolean `ready` field/accessor which indicates the readiness of the provider.
```

File: specification/sections/02-providers.md
Comment (justinabrahms) [inline specification/sections/02-providers.md:199]: Should we write a corresponding MUST NOT in the case of failure?

File: specification/sections/01-flag-evaluation.md
Comment (lukas-reining) [reply specification/sections/01-flag-evaluation.md:38]: Looks good, this makes it clearer🙂

File: specification/sections/02-providers.md
Comment (toddbaert) [reply specification/sections/02-providers.md]: I've thought about this and I think it's wise to change our `ready` boolean to an enum [as is often the case](https://wiki.c2.com/?UseEnumsNotBooleans).

This way we can give more information to users (especially integrators). There's also a distinct difference between the states `I've yet to be initialized` and `I tried to initialize but experienced an error`. I don't think we can capture both of these in a `ready` boolean.

Events (again) are very complementary here. Transitions between `NOT_READY`/`READY`/`ERROR` are all probably useful events, and maintaining equivalent states seems appropriate. Thanks for your continued attention to detail. See [delta](https://github.com/open-feature/spec/pull/179/commits/8ac074681ecb4196d4ce20633a0ff987066a2465).

cc @beeme1mr

File: specification/sections/01-flag-evaluation.md
Comment (Kavindu-Dodan) [reply specification/sections/01-flag-evaluation.md:36]: I also see the benefit of **global events** and IMO is the right way to support events. 

When an application starts, there is a setup phase where runtime configurations are set. I see Provider registration to be part of such a setup phase. And the client can be derived at business logic which can happen after some time (seconds to minutes depending on the usage) of the startup/setup phase. And it could be too late to detect events/failures at this point. Consider a K8s app with a startup probe. A developer can decide readiness to be set only if provider initialization is successful (aka `PROVIDER_READY`).

File: specification/sections/02-providers.md
Comment (Kavindu-Dodan) [inline specification/sections/02-providers.md:193]: Optional, may be a mermaid diagram to visualize the state transition 

```mermaid
---
title: Provider State
---
stateDiagram-v2
    direction LR
    [*] --> NOT_READY
    NOT_READY --> READY
    READY --> ERROR
    ERROR --> READY
```

Comment (Kavindu-Dodan) [reply specification/sections/02-providers.md:193]: And IMO these states needs to be added to `types.md` & later connect to provider events [1] introduced at  https://github.com/open-feature/spec/pull/182 

[1] - https://github.com/open-feature/spec/blob/23c6c229d9202e3a532b595e7eb80e1476334481/specification/types.md#provider-events

File: specification/sections/02-providers.md
Comment (Kavindu-Dodan) [reply specification/sections/02-providers.md]: Agree with this change (also see comment [1]). This change helps to handle error state which was not supported previously and help to differentiate not ready vs failure clearly. 

[1] - https://github.com/open-feature/spec/pull/179#discussion_r1184382057

File: specification/sections/02-providers.md
Comment (toddbaert) [reply specification/sections/02-providers.md:193]: I love this addition.

Comment (toddbaert) [reply specification/sections/02-providers.md:193]: [Thanks](https://github.com/open-feature/spec/pull/179/commits/6ebf5870de5ddc081d2ee3ad3e14a44f4a9670b6).

File: specification/types.md
Comment (lukas-reining) [inline specification/types.md:105]: I think we have a typo here with "initialed".

File: specification/sections/01-flag-evaluation.md
Comment (lukas-reining) [reply specification/sections/01-flag-evaluation.md:36]: _Not sure to which discussion to put this:_
Requirement 2.4.4 describes that transient errors can lead to the status flag changing between `ERROR` and `READY`. 
Should we specify that the change of the status Flag **MUST** lead to an event dispatched? (E.g. PROVIDER_READY and PROVIDER_ERROR)
Because if we only specify that there must be a (global) event on an error in `initialize()`, application integrators and application authors would get an inconsistent view on the state if the error is transient and status changes back to ready.

File: specification/types.md
Comment (jonathannorris) [reply specification/types.md:105]:
```suggestion
| ERROR     | The provider is initialized but is not able to reliably resolve flag values.      |
```

File: specification/sections/02-providers.md
Comment (jonathannorris) [inline specification/sections/02-providers.md:179]: Does it make sense for `initialize()` to potentially have an optional Promise / Callback mechanism to set the provider's state? That would allow you to add an optional Promise / Callback based initialization to the OpenFeature side allowing customers to require initialization before any flags are evaluated. In some server-side implementations waiting for initialization is preferable to serving defaults before the provider is ready.

Comment (toddbaert) [reply specification/sections/02-providers.md:179]: You're absolutely right. You may have heard me mention that there's a bit of coupling between the init/shutdown features and [events](https://github.com/open-feature/spec/pull/182). In the events spec changes, we're defining means to subscribe to events, one of them being provider readiness.

File: specification/sections/01-flag-evaluation.md
Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: > _Not sure to which discussion to put this:_ Requirement 2.4.4 describes that transient errors can lead to the status flag changing between `ERROR` and `READY`. Should we specify that the change of the status Flag **MUST** lead to an event dispatched? (E.g. PROVIDER_READY and PROVIDER_ERROR) Because if we only specify that there must be a (global) event on an error in `initialize()`, application integrators and application authors would get an inconsistent view on the state if the error is transient and status changes back to ready.

I agree in principle, but I think we'd have to put that in the events spec. Like I said in slack, these subjects are becoming a bit intertwined, but I'd like to keep the PRs isolated. I think when we release a new spec version, it should contain both init/shudown and events.

Comment (lukas-reining) [reply specification/sections/01-flag-evaluation.md:36]: Okay makes sense, so if we still want to include that, should it be added to the other PR (which is then the same problem, or what do we do to have it in the spec later @toddbaert?

Comment (toddbaert) [reply specification/sections/01-flag-evaluation.md:36]: > Okay makes sense, so if we still want to include that, should it be added to the other PR (which is then the same problem that we are talking about something not present there), or what do we do to have it in the spec later @toddbaert? Because it also feels kind of wrong to add it to that PR and talk about `initialize()` which is not present there.

`Initiailize()` is present there, that PR is based on this one. I'm just trying to keep discussion scoped - the feedback we've had before is to try to keep PRs as on-topic as possible

**Review by lukas-reining — APPROVED** [review]
Looks good to me now! 🙂

**Pull request merged**
